program/src/processor.rs

```diff
@@ -601,83 +658,127 @@ pub fn process_withdraw_inactive_stake(
         return Err(LidoError::ValidatorBalanceDecreased.into());
     }
 
     // We tracked in `stake_accounts_balance` what we put in there ourselves, so
     // the excess is a donation by some joker.
-    let donation = (observed_total - validator.entry.stake_accounts_balance)
+    let donation = (stake_observed_total - validator.entry.effective_stake_balance())
         .expect("Does not underflow because observed_total >= stake_accounts_balance.");
     msg!("{} in donations observed.", donation);
 
+    // Try to withdraw from unstake accounts.
+    let mut unstake_removed = Lamports(0);
+    let mut unstake_observed_total = Lamports(0);
+    for (i, seed) in validator.entry.unstake_seeds.into_iter().enumerate() {
+        let (unstake_account_address, _bump_seed) =
+            validator.find_unstake_account_address(program_id, accounts.lido.key, seed);
+
+        let unstake_account = &unstake_accounts[i];
+        let account_balance = get_stake_account_balance(
+            &unstake_account_address,
+            unstake_account,
+            seed,
+            StakeType::Unstake,
+        )?;
+
+        // If validator's stake is at the beginning, try to withdraw.
+        if validator.entry.unstake_seeds.begin == seed {
+            let unstake_removed_here = withdraw_inactive_sol(
+                &WithdrawExcessOpts {
+                    accounts: &accounts,
+                    clock: &clock,
+                    rent: &rent,
+                    stake_history: &stake_history,
+                    stake_account: unstake_account,
+                },
+                seed,
+                lido.stake_authority_bump_seed,
+                StakeType::Unstake,
+            )?;
+            assert_eq!(account_balance, unstake_removed_here);
+            validator.entry.unstake_seeds.begin += 1;
+            unstake_removed = (unstake_removed + unstake_removed_here)?;
+        }
+        unstake_observed_total = (unstake_observed_total + account_balance)?;
+    }
+
+    if unstake_observed_total < validator.entry.unstake_accounts_balance {
+        msg!(
+            "Observed balance of {} is less than tracked balance of {}.",
+            unstake_observed_total,
+            validator.entry.unstake_accounts_balance
+        );
+        msg!("This should not happen, aborting ...");
+        return Err(LidoError::ValidatorBalanceDecreased.into());
+    }
+
     // Store the new total. If we withdrew any inactive stake back to the
     // reserve, that is now no longer part of the stake accounts, so subtract
-    // that.
-    validator.entry.stake_accounts_balance =
-        (observed_total - excess_removed).expect("Does not underflow, because excess <= total.");
+    // that + the total unstake removed.
+    validator.entry.stake_accounts_balance = ((validator.entry.stake_accounts_balance
+        - excess_removed)
+        .expect("Does not underflow, because excess <= total.")
+        - unstake_removed)
```

Comment:
It’s a superficial thing, but with a chain of arithmetic and interspersed `expect`s like this, I find it nicer to use the method versions, because they `.`-chain nicely:

```rust
validator.entry.stake_accounts_balance = stake_observed_total
    .add(unstake_observed_total)
    .expect("If Solido has enough SOL to make this overflow, something has gone very wrong.")
    .sub(excess_removed)
    .expect(...)
    .sub(unstake_removed)
    .expect(...);
```